File: NativeApp/ViewModels/DeepSeekViewModel.swift
import Foundation
import SwiftUI
import Combine
import AppKit

// MARK: - DeepSeek ViewModel
@MainActor
class DeepSeekViewModel: ObservableObject {
    @Published var apiKey: String {
        didSet {
            KeychainHelper.save(key: "DeepSeekAPIKey", data: apiKey)
        }
    }
    @Published var refreshInterval: Int {
        didSet {
            UserDefaults.standard.set(refreshInterval, forKey: "DeepSeekRefreshInterval")
            startAutoRefresh()
        }
    }
    @Published var balance: String = "--"
    @Published var currency: String = "USD"
    @Published var spentToday: String = "0.0000"
    @Published var isAvailable: Bool = true
    @Published var selectedModel: String = APIConstants.DeepSeek.defaultModel
    @Published var latencyMs: Int = 0
    @Published var isLoading: Bool = false
    @Published var isGenerating: Bool = false
    @Published var promptText: String = ""
    @Published var aiResponse: String = ""
    @Published var copySuccess: Bool = false
    
    private var cancellables = Set<AnyCancellable>()
    private var refreshTimer: Timer?
    
    init() {
        self.apiKey = KeychainHelper.load(key: "DeepSeekAPIKey") ?? ""
        let savedInterval = UserDefaults.standard.integer(forKey: "DeepSeekRefreshInterval")
        self.refreshInterval = savedInterval > 0 ? savedInterval : 60
        fetchData()
        startAutoRefresh()
        AppDelegate.shared?.updateDeepSeekTitle(balance: self.balance)
    }
    
    func startAutoRefresh() {
        refreshTimer?.invalidate()
        guard refreshInterval > 0 else { return }
        refreshTimer = Timer.scheduledTimer(withTimeInterval: TimeInterval(refreshInterval), repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.fetchData()
            }
        }
    }
    
    func fetchData() {
        guard !apiKey.isEmpty, let url = URL(string: APIConstants.DeepSeek.balanceURL) else {
            isAvailable = false
            isLoading = false
            return
        }
        isLoading = true
        let start = Date()
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        
        URLSession.shared.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: BalanceResponse.self, decoder: JSONDecoder())
            .receive(on: DispatchQueue.main)
            .sink(receiveCompletion: { [weak self] completion in
                self?.isLoading = false
                if case .failure = completion {
                    self?.isAvailable = false
                }
            }, receiveValue: { [weak self] res in
                guard let self = self else { return }
                self.isLoading = false
                self.latencyMs = Int(Date().timeIntervalSince(start) * 1000)
                self.isAvailable = res.isAvailable
                if let info = res.balanceInfos?.first {
                    self.balance = info.totalBalance
                    self.currency = info.currency
                    self.updateDailySpend(currentBalance: info.totalBalance)
                    AppDelegate.shared?.updateDeepSeekTitle(balance: info.totalBalance)
                    
                    // Low balance notification check
                    if let doubleBalance = Double(info.totalBalance), doubleBalance < APIConstants.DeepSeek.lowBalanceThreshold {
                        NotificationService.shared.sendNotification(
                            identifier: "low_deepseek_balance",
                            title: "⚠️ DeepSeek Balance Low",
                            body: "Your remaining DeepSeek balance is $\(info.totalBalance) \(info.currency). Please top up soon."
                        )
                    }
                }
            })
            .store(in: &cancellables)
    }
    
    private func updateDailySpend(currentBalance: String) {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let todayStr = formatter.string(from: Date())
        
        let savedDate = UserDefaults.standard.string(forKey: "DailySpendDate") ?? ""
        let startBalStr = UserDefaults.standard.string(forKey: "DailySpendStartBal") ?? currentBalance
        
        guard let current = Double(currentBalance) else { return }
        
        if savedDate == todayStr {
            if let start = Double(startBalStr) {
                if current > start {
                    UserDefaults.standard.set(currentBalance, forKey: "DailySpendStartBal")
                    self.spentToday = "0.0000"
                } else {
                    let diff = max(0, start - current)
                    self.spentToday = String(format: "%.4f", diff)
                }
            }
        } else {
            UserDefaults.standard.set(todayStr, forKey: "DailySpendDate")
            UserDefaults.standard.set(currentBalance, forKey: "DailySpendStartBal")
            self.spentToday = "0.0000"
        }
    }
    
    func sendPrompt(systemPrompt: String? = nil) {
        guard !promptText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, !apiKey.isEmpty else { return }
        guard let url = URL(string: APIConstants.DeepSeek.chatCompletionsURL) else { return }
        
        isGenerating = true
        let start = Date()
        
        var messages: [[String: String]] = []
        if let sys = systemPrompt {
            messages.append(["role": "system", "content": sys])
        }
        messages.append(["role": "user", "content": promptText])
        
        let body: [String: Any] = [
            "model": selectedModel,
            "messages": messages,
            "max_tokens": 1000
        ]
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { [weak self] data, _, _ in
            DispatchQueue.main.async {
                guard let self = self else { return }
                self.isGenerating = false
                self.latencyMs = Int(Date().timeIntervalSince(start) * 1000)
                if let data = data,
                   let decoded = try? JSONDecoder().decode(ChatResponse.self, from: data),
                   let content = decoded.choices?.first?.message.content {
                    self.aiResponse = content
                    self.promptText = ""
                    self.fetchData()
                } else {
                    self.aiResponse = "Error getting response from DeepSeek API."
                }
            }
        }.resume()
    }
    
    func copyToClipboard() {
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(aiResponse, forType: .string)
        copySuccess = true
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
            self.copySuccess = false
        }
    }
}

File: WindowsApp/AIWidgetWindowsApp/ViewModels/DeepSeekViewModel.cs
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Windows;
using AIWidgetWindowsApp.Helpers;
using AIWidgetWindowsApp.Models;
using AIWidgetWindowsApp.Services;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;

namespace AIWidgetWindowsApp.ViewModels
{
    public partial class DeepSeekViewModel : ObservableObject
    {
        private readonly HttpClient _httpClient = new HttpClient();
        
        [ObservableProperty]
        private string apiKey = string.Empty;

        partial void OnApiKeyChanged(string value)
        {
            KeychainHelper.Save("DeepSeekAPIKey", value);
        }

        private System.Windows.Threading.DispatcherTimer? _refreshTimer;

        [ObservableProperty]
        private int refreshInterval = 60;

        partial void OnRefreshIntervalChanged(int value)
        {
            KeychainHelper.Save("DeepSeekRefreshInterval", value.ToString());
            StartAutoRefresh();
        }

        [ObservableProperty]
        private string balance = "--";

        [ObservableProperty]
        private string currency = "USD";

        [ObservableProperty]
        private string spentToday = "0.0000";

        [ObservableProperty]
        private bool isAvailable = true;

        [ObservableProperty]
        private string selectedModel = APIConstants.DeepSeek.DefaultModel;

        [ObservableProperty]
        private int latencyMs = 0;

        [ObservableProperty]
        private bool isLoading = false;

        [ObservableProperty]
        private bool isGenerating = false;

        [ObservableProperty]
        private string promptText = string.Empty;

        [ObservableProperty]
        private string aiResponse = string.Empty;

        [ObservableProperty]
        private bool copySuccess = false;

        public DeepSeekViewModel()
        {
            this.ApiKey = KeychainHelper.Load("DeepSeekAPIKey") ?? string.Empty;
            
            if (int.TryParse(KeychainHelper.Load("DeepSeekRefreshInterval"), out int savedInterval) && savedInterval > 0)
            {
                this.RefreshInterval = savedInterval;
            }
            
            _ = FetchDataAsync();
            StartAutoRefresh();
        }

        public void StartAutoRefresh()
        {
            _refreshTimer?.Stop();
            if (RefreshInterval > 0)
            {
                _refreshTimer = new System.Windows.Threading.DispatcherTimer();
                _refreshTimer.Interval = TimeSpan.FromSeconds(RefreshInterval);
                _refreshTimer.Tick += (s, e) => _ = FetchDataAsync();
                _refreshTimer.Start();
            }
        }

        private void UpdateDailySpend(string currentBalance)
        {
            string todayStr = DateTime.Now.ToString("yyyy-MM-dd");
            string savedDate = KeychainHelper.Load("DailySpendDate") ?? "";
            string startBalStr = KeychainHelper.Load("DailySpendStartBal") ?? currentBalance;

            if (double.TryParse(currentBalance, out double current))
            {
                if (savedDate == todayStr)
                {
                    if (double.TryParse(startBalStr, out double start))
                    {
                        if (current > start)
                        {
                            KeychainHelper.Save("DailySpendStartBal", currentBalance);
                            SpentToday = "0.0000";
                        }
                        else
                        {
                            double diff = Math.Max(0, start - current);
                            SpentToday = diff.ToString("F4");
                        }
                    }
                }
                else
                {
                    KeychainHelper.Save("DailySpendDate", todayStr);
                    KeychainHelper.Save("DailySpendStartBal", currentBalance);
                    SpentToday = "0.0000";
                }
            }
        }

        [RelayCommand]
        public async Task FetchDataAsync()
        {
            if (string.IsNullOrWhiteSpace(ApiKey))
            {
                IsAvailable = false;
                return;
            }

            IsLoading = true;
            var stopwatch = Stopwatch.StartNew();

            try
            {
                var request = new HttpRequestMessage(HttpMethod.Get, APIConstants.DeepSeek.BalanceURL);
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", ApiKey);

                var response = await _httpClient.SendAsync(request);
                stopwatch.Stop();
                LatencyMs = (int)stopwatch.ElapsedMilliseconds;

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    var balanceResponse = JsonSerializer.Deserialize<BalanceResponse>(json);

                    if (balanceResponse != null)
                    {
                        IsAvailable = balanceResponse.IsAvailable;
                        if (balanceResponse.BalanceInfos?.Count > 0)
                        {
                            var info = balanceResponse.BalanceInfos[0];
                            Balance = info.TotalBalance ?? "--";
                            Currency = info.Currency ?? "USD";
                            
                            UpdateDailySpend(Balance);
                            UpdateMainWindow($"${Balance} {Currency}");

                            if (double.TryParse(Balance, out double doubleBalance) && doubleBalance < APIConstants.DeepSeek.LowBalanceThreshold)
                            {
                                NotificationService.SendNotification("low_deepseek_balance", "⚠️ DeepSeek Balance Low", $"Your remaining DeepSeek balance is ${Balance} {Currency}. Please top up soon.");
                            }
                        }
                    }
                }
                else
                {
                    IsAvailable = false;
                }
            }
            catch (Exception)
            {
                IsAvailable = false;
            }
            finally
            {
                IsLoading = false;
            }
        }

        [RelayCommand]
        public async Task SendPromptAsync(string? systemPrompt = null)
        {
            if (string.IsNullOrWhiteSpace(PromptText) || string.IsNullOrWhiteSpace(ApiKey)) return;

            IsGenerating = true;
            var stopwatch = Stopwatch.StartNew();

            try
            {
                var request = new HttpRequestMessage(HttpMethod.Post, APIConstants.DeepSeek.ChatCompletionsURL);
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", ApiKey);

                var messages = new System.Collections.Generic.List<object>();
                if (!string.IsNullOrWhiteSpace(systemPrompt))
                {
                    messages.Add(new { role = "system", content = systemPrompt });
                }
                messages.Add(new { role = "user", content = PromptText });

                var body = new
                {
                    model = SelectedModel,
                    messages = messages,
                    max_tokens = 1000
                };

                request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

                var response = await _httpClient.SendAsync(request);
                stopwatch.Stop();
                LatencyMs = (int)stopwatch.ElapsedMilliseconds;

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    var chatResponse = JsonSerializer.Deserialize<ChatResponse>(json);
                    
                    if (chatResponse?.Choices?.Count > 0)
                    {
                        AiResponse = chatResponse.Choices[0].Message?.Content ?? "";
                        PromptText = string.Empty;
                        _ = FetchDataAsync();
                    }
                }
                else
                {
                    AiResponse = "Error getting response from DeepSeek API.";
                }
            }
            catch (Exception ex)
            {
                AiResponse = $"Error: {ex.Message}";
            }
            finally
            {
                IsGenerating = false;
            }
        }

        [RelayCommand]
        public async Task CopyToClipboardAsync()
        {
            Clipboard.SetText(AiResponse);
            CopySuccess = true;
            await Task.Delay(2000);
            CopySuccess = false;
        }

        private void UpdateMainWindow(string value)
        {
            if (Application.Current.MainWindow is MainWindow mainWindow)
            {
                mainWindow.UpdateTrayTextIfSelected(1, value);
            }
        }
    }
}
